src/librustc/ty/subst.rs

```diff
@@ -298,20 +275,20 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
                        target_substs: &Substs<'tcx>)
                        -> &'tcx Substs<'tcx> {
         let defs = tcx.lookup_generics(source_ancestor);
-        Substs::new(tcx, target_substs.iter().chain(&self[defs.own_count()..]).cloned())
+        tcx.mk_substs(target_substs.iter().chain(&self[defs.own_count()..]).cloned())
```

Comment:
We want to avoid `Iterator::cloned()`, but the types don't match up here to allow that, since we're iterating over `&Substs`, not `Substs`. I'm not sure if there is/what the good solution here is; I believe this is semi-common though.